Wallet Decrypt - Mnemonic

skubakdj opened the pull request:
### **What This Does**
Broadly speaking, two things are accomplished: 
- Adds mnemonic wallet decrypt as described in #103 & referenced by #80 
- Adds mnemonic support to the `DeterministicWalletsModal` & accompanying sagas

### **What Still Needs to Happen**
A few things to get done before this is merged:
- ~~Need to write unit tests for `decryptMnemonicToPrivKey`~~
- ~~Need to do more manual testing~~ 
- Need to test against Trezor to make sure I didn't break anything (may need some help for this one)

skubakdj commented on common/libs/validators.js:
```diff
@@ -143,5 +143,7 @@ export function isValidRawTx(rawTx: RawTransaction): boolean {
-  return dPath.split("'/").length === 4;
+  //TODO: use a regex to detect proper paths
+  const len = dPath.split("'/").length;
+  return len === 3 || len === 4;
```
Ledger uses `m/44'/60'/0'` for some reason:

https://ethereum.stackexchange.com/questions/19055/what-is-the-difference-between-m-44-60-0-0-and-m-44-60-0/19061

wbobeirne replied:
Oh wow, BIP44 is way more readable than BIP32. Would have saved me a lot of headaches: https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki

Would you mind changing the comment to reflect that BIP 44 is really where we're determining that 4 is the amount, but that ledger is the exception for 3?

skubakdj replied:
yep, will do :+1:

skubakdj commented on common/sagas/deterministicWallets.js:
```diff
@@ -62,16 +72,22 @@ function* getDeterministicWallets(
-      })
-    );
+
+  try {
```
`try/catch` introduced because derived addresses wouldn't populate on network/api failure

dternyak commented on common/components/WalletDecrypt/Mnemonic.jsx:
```diff
@@ -1,27 +1,129 @@
+                : 'is-invalid'}`}
+              value={phrase}
+              onChange={this.onMnemonicChange}
+              placeholder={translateRaw('x_Mnemonic')}
```
You can save the `translateRaw` import by passing `true` as the argument for the `textOnly` parameter:

`translate('x_Mnemonic', true)`

dternyak replied:
Makes me question if `translateRaw` should even be exported.

wbobeirne commented on common/config/dpaths.json:
```diff
@@ -12,5 +12,31 @@
+    {
+      "label": "Expanse",
+      "value": "m/44'/40'/0'/0"
+    }
```
Seeing multiple now, might be worth making this a JS file so that we can assign these repeated ones to variables and reuse them.

skubakdj replied:
Converted `dpaths.json` to `dpaths.js`.

wbobeirne commented on common/components/WalletDecrypt/DeterministicWalletsModal.jsx:
```diff
@@ -40,13 +40,15 @@ type Props = {
   dPaths: { label: string, value: string }[],
   publicKey: string,
   chainCode: string,
+  seed: string,
```
This is a required prop, but the DecryptWallet/Trezor.jsx wasn't altered to send it (And probably shouldn't, I think this should be optional.) Any idea why flow didn't catch that?

skubakdj replied:
Agree, I'll make it optional. Not sure why flow didn't catch that, it definitely should have...

wbobeirne commented on common/sagas/deterministicWallets.js:
```diff
@@ -62,16 +72,22 @@ function* getDeterministicWallets(
+    }
+  } catch (err) {
+    console.log(err);
+    //TODO: maybe communicate API call error to user?
```
Definitely, I think you can just get away with a `yield put(showNotification(...`

wbobeirne commented on common/sagas/deterministicWallets.js:
```diff
@@ -86,21 +102,27 @@ function* updateWalletTokenValues(): Generator<Yield, Return, Next> {
+    }
+  } catch (err) {
+    console.log(err);
+    //TODO: maybe communicate API call error to user?
```
Same as above, showNotification

skubakdj commented on common/components/WalletDecrypt/Mnemonic.jsx:
```diff
@@ -1,27 +1,129 @@
+    this.setState({ phrase: e.target.value });
+  };
+
+  onChooseAddress = (e: SyntheticInputEvent) => {
```
Good call, I'll switch it to `onDWModalOpen`, unless you can suggest something else.

wbobeirne commented:
Tested the branch on my Trezor, everything seems to work fine. Still not sure how it's not complaining more about the lack of seed arg, but looks like no regressions here.

dternyak commented:
@wbobeirne Interesting. I'm currently unable to send via Trezor on the latest develop. You're able to send via Trezor on this branch? Notice: The signed tx generates, but when attempting to broadcast it the node always reponds with 
```
Error: Insufficient funds. The account you tried to send transaction from does not have enough funds. Required XXX000000000000 and got: 0.
```

dternyak commented:
I'm going to attempt to send via Trezor on this branch -- will update with my results.

wbobeirne commented:
Sorry, allow me to reiterate, I only tested unlocking. This branch is a bit behind develop, and I didn't test sending until it was up to date.

wbobeirne commented on common/components/WalletDecrypt/index.jsx:
```diff
@@ -165,7 +171,7 @@ export class WalletDecrypt extends Component {
   onUnlock = (payload: any) => {
     this.props.dispatch(
-      WALLETS[this.state.selectedWalletKey].unlock(this.state.value || payload)
+      WALLETS[this.state.selectedWalletKey].unlock(payload || this.state.value)
```
Payload is a SyntheticEvent in some cases. We'll either need to check if payload isn't one, or change how onUnlock is called so that it's never sent an event.

dternyak replied:
I believe I have fixed this in one of my recent PRs https://github.com/MyEtherWallet/MyEtherWallet/pull/183/files#diff-c2a1968700ed6452c7c0257d4dde4cd4R168

wbobeirne replied:
Sweet, in that case we just need to bring `develop` back into this bad boy.

skubakdj commented:
@dternyak, @wbobeirne I think I've addressed all the comments with these latest commits. Also added some tests, rebased this branch with `develop`, and did some manual testing. Any additional feedback on the changes?

Also, still haven't been able to test using Trezor, but I can get to it tonight if need be.

dternyak commented:
@wboberine is working on a fix to Trezor, and is planning on commiting that
directly to this branch.

After that we should be good to merge! :)


On Thu, Sep 14, 2017 at 11:13 AM skubakdj <notifications@github.com> wrote:

> @dternyak <https://github.com/dternyak>, @wbobeirne
> <https://github.com/wbobeirne> I think I've addressed all the comments
> with these latest commits. Also added some tests, rebased this branch with
> develop, and did some manual testing. Any additional feedback on the
> changes?
>
> Also, still haven't been able to test using Trezor, but I can get to it
> tonight if need be.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/MyEtherWallet/MyEtherWallet/pull/180#issuecomment-329565244>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHf02TBDYLVBm_EEf_g_q521wzHQ9Hsuks5siWy-gaJpZM4PQtnT>
> .
>
